File: backend/src/dispatcher/network-connection.ts
import { PublicMetaInfo } from "./protocol/meta";
import { TaskResult } from "./protocol/task";
import { Task, TaskStatus } from "./task";
import { NOOP } from "./utils/noop";


// TODO: Refactor to interface
export class NetworkConnection {
  async sendTask(task: Task): Promise<void> {
    // do await
    task.setStatus(TaskStatus.SentToProvider);
  }

  async abortTask(task: Task): Promise<void> {
    task.setStatus(TaskStatus.Aborted);
    // do await
  }

  async close(): Promise<void> {
    // drop websocket connection
  }

  onMetaUpdated: (meta: PublicMetaInfo) => void = NOOP;

  onLostConnection: () => void = NOOP;

  onConnectionRestored: () => void = NOOP;

  onTaskCompleted: (task: Task, result: TaskResult) => void = NOOP;

  onTaskFailed: (task: Task, reason: string) => void = NOOP;

  // TODO:
  // onTaskTimeout: (task: Task) => void = NOOP;

  // TODO: onTaskRejected, onTaskAccepted, onProgressUpdate
}

File: backend/src/dispatcher/network_connection.py
from typing import Callable
from dispatcher.protocol.meta import PublicMetaInfo
from dispatcher.protocol.task import TaskResult, Task, TaskStatus
from dispatcher.utils.noop import NOOP

class NetworkConnection:
    async def sendTask(self, task: Task) -> None:
        await task.setStatus(TaskStatus.SentToProvider)
  
    async def abortTask(self, task: Task) -> None:
        task.setStatus(TaskStatus.Aborted)
    
    async def close(self) -> None:
        pass
  
    onMetaUpdated: Callable[[PublicMetaInfo], None] = NOOP

    onLostConnection: Callable[[], None] = NOOP

    onConnectionRestored: Callable[[], None] = NOOP

    onTaskCompleted: Callable[[Task, TaskResult], None] = NOOP

    onTaskFailed: Callable[[Task, str], None] = NOOP

  # onTaskTimeout: Callable[[Task], None] = NOOP

  # onTaskRejected, onTaskAccepted, onProgressUpdate
